### PR TITLE
Blog draft

### PR DESCRIPTION
**Pull request by citerana**
PR for blog post review.

### CODE REVIEW

File: planet-kaggle-blog.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
+## The Planet Kaggle
+
+![](imgs/banner.jpg)
+
+### Amazon Basin Data
+
+![](imgs/chipdesc.jpg)
+<p align="center">source: <a href="https://www.kaggle.com/c/planet-understanding-the-amazon-from-space/data">Planet</a></p>
+
+This past summer, Planet launched the *Understanding the Amazon from Space* Kaggle competition. We participated in this competition using Raster Vision, a system for analyzing aerial and satellite imagery using deep learning. Raster Vision works across several different tasks, including semantic segmentation, object detection and scene tagging as well as a range of datasets. The varied data within the Amazon competition gave us a challenging opportunity to make a robust scene tagging functionality.
```
Comment (lewfish) [inline planet-kaggle-blog.md]: 'a challenging opportunity to make a robust scene tagging functionality' sounds awkward. I think 'a challenging opportunity to add scene tagging functionality' sounds better. Not sure the robust part is necessary.

File: planet-kaggle-blog.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+Planet provided over 100,000 chips from large images taken by a flock of satellites over the Amazon basin in 2016. These 40,000 training and 60,000 testing chips were given in both 3-band RGB JPEG and four band IR-RBG TIFF formats. The goal of tagging is to infer a set of labels for a given chip. For the Amazon images, there were 17 possible tags, which could be broadly split into three categories.
+
+1. **atmospheric labels**: clear, partly cloudy, cloudy and hazy
+2. **common labels**: primary, water, habitation, agriculture, road, cultivation and bare ground
+3. **rare labels**: artisinal mine, blooming, blow down, conventional mine, selective logging and slash burn.
+
+Each of the chips were labeled with ground truth tags through crowd-sourced labor. The ground truth is used to train models on the correct labels for a given chip. A well trained network can accurately predict the labels on a previously unseen chip.
```
Comment (lewfish) [inline planet-kaggle-blog.md]: 'The ground truth is used to train models on the correct labels for a given chip.' Since 'ground truth' is synonymous with 'correct labels', this sentence seems kind of tautological. Also, we should pick either tags or labels to be consistent through the article.

File: planet-kaggle-blog.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+<p align="center"><img src="imgs/tags_hist.png" alt="Tag histogram"/></p>
+<p align="center">source: <a href="https://www.kaggle.com/anokas/data-exploration-analysis">anokas</a></p>
+
+For this dataset, the tags were quite varied in frequency. Primary was by far the most common tag, appearing in nearly 39,000 of the provided chips. Rare labels, on the other hand, were a source of concern. There were so few samples of rare labels, like conventional mining and blow down, that after splitting a portion of training data into a validation hold-out set, it was possible to have less than 100 data points for some rare tags versus the thousands of examples of images containing primary rainforest. Without teaching the model to pay more attention to rarer labels, uncommon features would be difficult to learn and commonly missed or wrongly predicted by the model.
```
Comment (lewfish) [inline planet-kaggle-blog.md]: 'Without teaching the model to pay more attention to rarer labels, uncommon features would be difficult to learn and commonly missed or wrongly predicted by the model.' We tried this (using oversampling) but it didn't actually help.

File: planet-kaggle-blog.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+### Our Approach
+
+Here are some examples of training chips with labels predicted by our best single-model network.
+
+![Example tagging](imgs/debug_plots_labeled.png)
+<p align="center">source: <a href="https://github.com/azavea/raster-vision/">Raster Vision</a></p>
```
Comment (lewfish) [inline planet-kaggle-blog.md:30]: This figure uses red and green text which is a problem for people who are color blind. I think the green text is also bolded, so maybe that's all the signal you need, but it would be better to change this to a different color combo.

File: planet-kaggle-blog.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+In the above figure, the ground truth tags (ie. tagged by hand) for the Planet Kaggle dataset are bolded. Green bolded tags are correct. Unbolded and uncolored tags mean that they have been incorrectly predicted for the chip. Red bolded tags are ones missed by the network prediction. As with many Kaggle competitions involving image identifcation, there are multiple stages in the prediction process that can be optimized for better accuracy. Improvements can be made in pre-processing the training data, switching or combining model architectures, adjusting optimizers, learning rates and augmenting the testing data. We placed 23rd overall out of nearly a thousand teams with a private leaderboard prediction accuracy of 93.154% using the following techniques.
```
Comment (lewfish) [inline planet-kaggle-blog.md]: For people familiar with ML, the terminology 'false positive' and 'false negative' is standard and easily understood, so you might want to add that, in addition to the phrasing you already have which is also useful. I mean to add this to the text, not the figure.

Comment (lewfish) [inline planet-kaggle-blog.md]: 'image identification' -> 'image analysis'

Comment (lewfish) [reply planet-kaggle-blog.md]: 'a private leaderboard prediction accuracy of 93.154%' -- actually, it was F2 score (which is not a percentage), not accuracy which is a different thing.

File: planet-kaggle-blog.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+We normalized the provided images to be within a standardized range and used a range of image augmentations during the testing phase. Rotation, zooming in/out, flipping and cropping did increase the quality of prediction in some cases. We used a multi-architecture ensemble that took a majority vote over the predicted labels across 15 convolutional neural networks: (5) [ResNet50](https://arxiv.org/abs/1512.03385), (5) [Inception v3](https://arxiv.org/abs/1512.00567) and (5) [DenseNet121](https://arxiv.org/abs/1608.06993). These models were all trained using Adam as the optimizer, a cyclically decaying learning rate and binary cross-entropy as the loss function.
```
Comment (lewfish) [inline planet-kaggle-blog.md]: 'Rotation, zooming in/out, flipping and cropping did increase the quality of prediction in some cases'. I think that zooming and cropping didn't actually help. I'm not sure what cases you're referring to.

Comment (lewfish) [inline planet-kaggle-blog.md]: 'cyclically decaying learning rate' I don't think we ever got this to yield an improvement in the score. We just dropped the learning rate twice after a certain number of epochs. This would be a good thing to discuss more along with a graph showing how the score changes after dropping the learning rate.

File: planet-kaggle-blog.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+The winning Kaggler, bestfitting, obtained a score of 93.318% accuracy with their solution. More details on this approach can be found in bestfitting's [solution summary](https://www.kaggle.com/c/planet-understanding-the-amazon-from-space/discussion/36809).
```
Comment (lewfish) [inline planet-kaggle-blog.md]: 'score of 93.318% accuracy' same comment as above

File: planet-kaggle-blog.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+
+A point of note: submissions to the competition were ranked using F2 score.  The F2 scoring metric tended to overlook mislabeling in underrepresented tags. Many high-performing trained their networks to improve the F2 score, we decided to not use this method. If we trained our model according to this metric, there would be less incentive for correctly identifying the rarer labels. Our method was conducted with the premise of fairly differentiating and labeling the uncommon tags. Often, the goal in using machine tagging is to teach the model to effectively and accurately identify interesting and anomalous features in a set of largely similar data.
```
Comment (lewfish) [inline planet-kaggle-blog.md]: 'The F2 scoring metric tended to overlook mislabeling in underrepresented tags.' True, but only due to the averaging strategy they used, which was to take the F2 score of each sample and then average over them, instead of taking the F2 score for each tag and average over them, which would have made more sense. This is something you should discuss.

Comment (lewfish) [reply planet-kaggle-blog.md]: 'Many high-performing trained their networks to improve the F2 score, we decided to not use this method.' I'm not sure what you're referring to here. We tuned the decision thresholds to optimize the F2 score too.

Comment (lewfish) [reply planet-kaggle-blog.md]: 'Our method was conducted with the premise of fairly differentiating and labeling the uncommon tags. Often, the goal in using machine tagging is to teach the model to effectively and accurately identify interesting and anomalous features in a set of largely similar data.' I'm having trouble understanding these two sentences.

File: planet-kaggle-blog.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+### Misconfigured Labels
```
Comment (lewfish) [inline planet-kaggle-blog.md]: Consider changing 'misconfigured' to 'incorrect' or 'mistaken'

File: planet-kaggle-blog.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+In order for this task to be performed accurately, it is crucial that the ground truth be genuinely truthful. In an ideal world, the criteria for a label is clear and distinct to humans and similarly obvious to trained neural networks. Unfortunately, Planet's crowd-labeled dataset had noticeable amounts of ambiguous and, even worse, clearly incorrect labels. For example, we can see that the network often mistakes when to label a chip with the tag `agriculture`. However, if we examine the ground truth tags for each chip, it's not obvious that the human classifications are correct either.
```
Comment (lewfish) [inline planet-kaggle-blog.md]: 'In order for this task to be performed accurately, it is crucial that the ground truth be genuinely truthful.' The ground truth is reality, so it's always correct by definition. I think you mean to say something more like 'the human-provided labels are mostly correct.'

Comment (lewfish) [inline planet-kaggle-blog.md]: 'Unfortunately, Planet's crowd-labeled dataset had noticeable amounts of ambiguous and, even worse, clearly incorrect labels.' I think 'many' would be a good replacement for 'noticeable amounts of'.

File: planet-kaggle-blog.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+Planet provided over 100,000 chips from large images taken by a flock of satellites over the Amazon basin in 2016. These 40,000 training and 60,000 testing chips were given in both 3-band RGB JPEG and four band IR-RBG TIFF formats. The goal of tagging is to infer a set of labels for a given chip. For the Amazon images, there were 17 possible tags, which could be broadly split into three categories.
+
+1. **atmospheric labels**: clear, partly cloudy, cloudy and hazy
+2. **common labels**: primary, water, habitation, agriculture, road, cultivation and bare ground
+3. **rare labels**: artisinal mine, blooming, blow down, conventional mine, selective logging and slash burn.
+
+Each of the chips were labeled with ground truth tags through crowd-sourced labor. The ground truth is used to train models on the correct labels for a given chip. A well trained network can accurately predict the labels on a previously unseen chip.
+
+<p align="center"><img src="imgs/tags_hist.png" alt="Tag histogram"/></p>
```
Comment (lewfish) [inline planet-kaggle-blog.md]: Typically histograms like this are sorted by frequency (most frequent bars to least) which makes it easier to parse. I remember seeing the same graph, except sorted on Planet's Jupyter notebook.

File: planet-kaggle-blog.md [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+### Discussion
+
+Many techniques that we implemented in Raster Vision and trialed did not increase our F2 score on the Amazon dataset. In several cases, we believe that the F2 score metric had unintentional effects on what constituted an accurate result. For example, we tried creating separate networks with different output layers for atmospheric tags versus the other tags. As there can be only one correct atmospheric tag, it seemed that a sigmoid layer would output better predictions than a softmax layer. Although counterintuitive to our expectation, this did not yield improved results. The metric places greater value on predicions with both correct and incorrect labels more than predictions which do not have the correct label at all. We also had experiments that artificially boosted the presence of rarer labels during training. We suspect that the F2 score simply did not place as much value on the performance for infrequent tags. Even if our predictions were more accurate for rare tags, it came at the cost of our overall F2 score across all the tags.
```
Comment (lewfish) [inline planet-kaggle-blog.md]: 'Many techniques that we implemented in Raster Vision and trialed' -> 'Many techniques we tried'

Comment (lewfish) [reply planet-kaggle-blog.md]: 'As there can be only one correct atmospheric tag, it seemed that a sigmoid layer would output better predictions than a softmax layer.' It's the other way around -- a softmax layer picks a single tag out of a set.

Comment (lewfish) [reply planet-kaggle-blog.md]: 'We also had experiments that artificially boosted the presence of rarer labels during training.' I don't think this helped, even when looking at the individual tag scores.

File: planet-kaggle-blog.md [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+
+Kaggle competitions are often won by very slim margins. In the case of the Amazon basin data, we spent a large portion of our development time trialing different methods of preprocessing and postprocessing the images. Kaggle competitions incentize improving your position on the public leaderboard. The score for the public leaderboard was calculated using 66% of the testing chips. The winner, however, is determined by the final private leaderboard score. This score uses the remaining 34% of the testing data and requires that the model appropriately combat overfitting to the public testing data too much. Since the private leaderboard performance is a blackbox until the competition deadline, generalizing the model is crucial to a high Kaggle rank. When the model approaches an accuracy that is hampered by the noise of the dataset itself, progression towards an incrementally better score can quickly become time, resource and computationally expensive. Many image processing, architectures or optmizers that did not improve our performance on the Amazon dataset remain useful for the scene tagging task and within Raster Vision.
```
Comment (lewfish) [inline planet-kaggle-blog.md]: 'postprocessing the images' -- which postprocessing methods? I think you mean postprocessing the predictions?

Comment (lewfish) [reply planet-kaggle-blog.md]: 'private leaderboard performance is a blackbox until the competition deadline' The term blackbox in engineering refers to something where you know the inputs and outputs, but don't know how the computation takes place. In that sense, I don't think the private leaderboard performance is a blackbox -- it's just unknown.

File: planet-kaggle-blog.md [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+
+### Future Work
+
+The open-source project repository for Raster Vision can be found [here](https://github.com/azavea/raster-vision/). We are currently developing and improving an accurate and scalable object detection functionality in Raster Vision, with a particular focus on strong performance for aerial and satellite imagery.
```
Comment (lewfish) [inline planet-kaggle-blog.md]: 'currently developing and improving an accurate and scalable object detection functionality in Raster Vision' -> 'currently incorporating object detection into Raster Vision'

File: planet-kaggle-blog.md (hunk 13 above)
Comment (lewfish) [reply planet-kaggle-blog.md]: 'When the model approaches an accuracy that is hampered by the noise of the dataset itself, progression towards an incrementally better score can quickly become time, resource and computationally expensive.' I'm a little confused by this sentence. I think you're trying to say that it takes great effort to make a small improvement in the score once you have a decent model.

**Review by lewfish — CHANGES_REQUESTED** [review]
I think you should talk more about: optimizing thresholds, talk about what F2 score is (an average of precision and recall that is weighted more towards recall), the different F2 average strategies and how they incentivize different things, more explanation of what an ensemble is and why it helps, our scores for a single best model vs ensemble, a learning curve and how long it took to train, that we used Keras and Tensorflow and how you liked them, the scores for each tag as opposed to the aggregate score (show with numbers what you alluded to). I think you could spend a bit less time talking about the problems with the data, and make the writing clearer and more concise along the lines of what I said in the individual comments.

File: planet-kaggle-blog.md (hunk 7 above)
Comment (lewfish) [inline planet-kaggle-blog.md]: Instead you should  link to this new blog post by the winner http://blog.kaggle.com/2017/10/17/planet-understanding-the-amazon-from-space-1st-place-winners-interview/
